src/chain/forkChoice/lmdGhost.ts

```diff
@@ -1,123 +1,252 @@
+    }
+  }
 }
 
 /**
- * Execute the LMD-GHOST algorithm to find the head BeaconBlock
- * @param {Store} store
- * @param {BeaconState} startState
- * @param {BeaconBlock} startBlock
- * @returns {BeaconBlock}
+ * Calculate best block using
+ * Latest Message-Driven Greedy Heaviest Observed SubTree
+ * See https://github.com/protolambda/lmd-ghost#state-ful-dag
  */
-export function lmdGhost(store: Store, startState: BeaconState, startBlock: BeaconBlock): BeaconBlock {
-  const validators = startState.validatorRegistry;
-  const activeValidatorIndices: ValidatorIndex[] = getActiveValidatorIndices(validators, slotToEpoch(startState.slot));
+export class LMDGHOST {
+  /**
+   * Aggregated attestations
+   */
+  private aggregator: AttestationAggregator;
+
+  /**
+   * Recently seen blocks, pruned up to last finalized block
+   */
+  private nodes: Record<Root, Node>;
+
+  /**
+   * Last finalized block
+   */
+  private finalized: Node;
 
-  const attestationTargets: AttestationTarget[] = [];
-  for (const validatorIndex of activeValidatorIndices) {
-    attestationTargets.push({validatorIndex, target: getLatestAttestationTarget(store, validatorIndex)});
+  /**
+   * Last justified block
+   */
+  private justified: Node;
+  private synced: boolean;
+
+  public constructor() {
+    this.aggregator = new AttestationAggregator((hex) => this.nodes[hex] ? this.nodes[hex].slot : null);
+    this.nodes = {};
+    this.finalized = null;
+    this.justified = null;
+    this.synced = true;
   }
 
-  // Inner function
-  function getVoteCount(block: BeaconBlock): int {
-    let sum = 0;
-    for (const target of attestationTargets) {
-      if (getAncestor(store, target[1], block.slot) === block) {
-        sum += Math.floor(getEffectiveBalance(startState, target[0]).toNumber() / FORK_CHOICE_BALANCE_INCREMENT);
-      }
+  public addBlock(slot: Slot, blockRootBuf: bytes32, parentRootBuf: bytes32): void {
+    this.synced = false;
+    const blockRoot = blockRootBuf.toString('hex');
+    const parentRoot = parentRootBuf.toString('hex');
+    // ensure blockRoot exists
+    const node: Node = this.nodes[blockRoot] || new Node({
+      slot,
+      blockRoot,
+      parent: this.nodes[parentRoot],
+    });
+    // best target is the node itself
+    node.bestTarget = node;
+    this.nodes[blockRoot] = node;
+
+    // if parent root exists, link to blockRoot
+    if (this.nodes[parentRoot]) {
+      this.nodes[parentRoot].addChild(node);
     }
-    return sum;
   }
 
-  let head = startBlock;
-  while (1) {
-    const children = getChildren(store, head);
-    if (children.length === 0) {
-      return head;
+  public addAttestation(blockRootBuf: bytes32, attester: ValidatorIndex, weight: Gwei): void {
+    this.synced = false;
+    this.aggregator.addAttestation({
+      target: blockRootBuf.toString('hex'),
+      attester,
+      weight,
+    });
+  }
+
+  public setFinalized(blockRoot: bytes32): void {
+    const rootHex = blockRoot.toString('hex');
```

Comment:
should `this.synced = false` here?

Reply:
I don't think so, but it won't hurt to add it. If there are no updates, the sync doesn't do anything.